src/herring/model.hpp

```diff
@@ -0,0 +1,275 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include<cstddef>
+#include<functional>
+#include<iostream>
+#include<new>
+#include<numeric>
+#include<type_traits>
+#include<vector>
+
+#include<herring/omp_helpers.hpp>
+#include<herring/output_ops.hpp>
+#include<herring/tree.hpp>
+#include<herring/type_helpers.hpp>
+
+
+
+namespace herring {
+
+#ifdef __cpp_lib_hardware_interference_size
+  using std::hardware_constructive_interference_size;
+#else
+  auto constexpr hardware_constructive_interference_size = std::size_t{64};
+#endif
+
+  template <typename value_t, typename feature_index_t, typename offset_t, typename output_index_t, typename output_t>
+  struct simple_model {
+    using simple_tree_t = simple_tree<value_t, feature_index_t, offset_t, output_index_t, output_t>;
+    using lookup_tree_t = lookup_tree<value_t, feature_index_t, offset_t, output_index_t, output_t>;
+    using tree_type = std::conditional_t<
+      std::is_same_v<value_t, output_t> || std::is_same_v<output_index_t, output_t>,
+      simple_tree_t,
+      lookup_tree_t
+    >;
+    using sum_elem_type = typename is_container_specialization<output_t, std::vector>::value_type;
+
+    std::vector<tree_type> trees;
+    std::size_t num_class;
+    std::size_t num_feature;
+    row_op row_postproc;
+    float average_factor;
+    float bias;
+    float postproc_constant;
+    std::vector<bool> mutable row_has_missing;
+    bool use_inclusive_threshold;
+
+    void predict(float const* input, std::size_t num_row, float* output, thread_count<int> nthread) const {
+      if (!precompute_missing(input, num_row)) {
+        if (!use_inclusive_threshold) {
+          predict_<false, false>(input, output, num_row, nthread);
+        } else {
+          predict_<false, true>(input, output, num_row, nthread);
+        }
+      } else {
+        if (!use_inclusive_threshold) {
+          predict_<true, false>(input, output, num_row, nthread);
+        } else {
+          predict_<true, true>(input, output, num_row, nthread);
+        }
+      }
+    }
+
+    void set_element_postproc(element_op element_postproc) {
+      postprocess_element = [this, element_postproc]() -> std::function<void(sum_elem_type, float*)> {
+        auto constant = postproc_constant;
+        switch(element_postproc) {
+          case element_op::signed_square:
+            return [](sum_elem_type elem, float* out) {
+              *out = std::copysign(elem * elem, elem);
+            };
```

Comment:
Any reason why you don't directly assign the lambda to `postprocess_element`?

Reply:
Nope! Good catch. Will fix it up.